### DISCUSSION TITLE
Proposal: Interprocedural analysis for NRT

### DISCUSSION THREAD

**Discussion by hez2010** [dotnet/csharplang#2792] [Language Ideas]
# Proposal
## Interprocedural analysis for NRT

A sample code:

```csharp
class A
{
        private string? x = null;
        private void OtherMethod()
        {
            x = "";
        }

        private void Test()
        {
            OtherMethod();
            var len = x.Length; // Dereference of a possible null value
        }
}
```

Currently compiler will throw "Dereference of a possible null value", however, with interprocedural analysis, `var len = x.Length` should not throw any NRT related issues.

**Comment by CyrusNajmabadi** [comment]
There is no interprocedural analysis in c#

My recommendation would be to have OriginalMethod return the non-null value

**Comment by hez2010** [comment]
@CyrusNajmabadi Is there any plan to support it? I think it can greatly improve the analysis ability of roslyn.

**Comment by CyrusNajmabadi** [comment]
not that i know of.  Def file over at dotnet/csharplang if you're interested in having this happen!

**Comment by hez2010** [comment]
I've update the issue and made a proposal. Any update on it?

**Comment by agocke** [comment]
I think this would be a big problem. Consider the following code:

```C#
class A
{
        internal string? x = null;
        private void OtherMethod()
        {
            x = "";
           SomeOtherMethod();
        }

        private void Test()
        {
            OtherMethod();
            var len = x.Length; // Dereference of a possible null value
        }
}
```

If SomeOtherMethod() is in metadata, what's the state of `x`? Maybe the proposal is only for private fields and private methods in the same class, but the value seems much more limited at that point, and this level of analysis could get very expensive, very quickly.

Even a slight change to your code:

```C#
class A
{
        private string? x = null;
        private void Test()
        {
            OtherMethod();
            var len = x.Length; // Dereference of a possible null value
        }
        private void OtherMethod()
        {
            x = "";
        }
}
```

This looks like it would at least double the time we need to analyze the class, since by the time we're analyzing `Test` we haven't analyzed `OtherMethod`, so we have to go back after we're done analyzing `OtherMethod` and do it all again, just in case the state changed.
